collections/binary_heap/src/min_heap.rs

```diff
@@ -0,0 +1,197 @@
+//!
+//! ## Sub Modules
+//!
+//! ### comparator
+//! Comparator trait의 구현체.
+//!
+
+use std::cmp::Ordering;
+use std::ops::{Deref, DerefMut};
+
+use crate::heap_logic::{build_heap, min_heapify, get_parent};
+use crate::comparator::{Comparator, DefaultComparator};
+
+pub struct MinHeap<T, C: Comparator<T>> {
+	data: Vec<T>,
+	comparator: C,
+}
+
+impl<T, C> MinHeap<T, C>
+where
+	C : Comparator<T>
+{
+	/// create empty min heap
+	pub fn new(comp : C) -> MinHeap<T, C> {
+		MinHeap {
+			data : Vec::new(),
+			comparator : comp,
+		}
+	}
+
+	/// create min heap with vector
+	pub fn from_vec(mut vec : Vec<T>, comp : C) -> MinHeap<T, C> {
+		let comparator = comp;
+		build_heap(&mut vec, &comparator);
+		MinHeap {
+			data : vec,
+			comparator,
+		}
+	}
+
+	/// push new element to the heap
+	pub fn push(&mut self, elem : T) {
+		let data = &mut self.data;
+		data.push(elem);
+		let mut cur_idx = data.len() - 1;
+		let mut parent_idx = get_parent(cur_idx);
+		while parent_idx < data.len()
+			&& cur_idx < data.len()
+			&& Ordering::Greater == self.comparator.compare(&data[parent_idx], &data[cur_idx])
+		{
+			data.swap(parent_idx, cur_idx);	// pull up
+			cur_idx = parent_idx;
+			parent_idx = get_parent(parent_idx);
+		}
+	}
+
+	/// add multiple element to the heap
+	pub fn extend(&mut self, elems : &mut Vec<T>) {
+		let data = &mut self.data;
+		data.append(elems);
+		build_heap(data,&self.comparator);	// O(n)
+	}
+
+	/// extract ownership of the element at root
+	/// O(log n)
+	pub fn pop(&mut self) -> Option<T> {
+		if self.is_empty() {
+			return None;
+		}
+		let data = &mut self.data;
+		let end_idx = data.len() - 1;	// already checked the emptiness, no overflow
+		data.swap(0, end_idx);
+		let result = data.pop();
+		min_heapify(data, &self.comparator, 0);	// O(log n)
+		result
+	}
+
+	/// # Description
+	/// get mutable reference of root of binary heap
+	/// it's source will be heaped when the PeekMut drops
+	pub fn peek_mut<'a>(&'a mut self) -> Option<PeekMut<'a, T, C>> {
+		if !self.is_empty() {
+			return Some(PeekMut {
+				comp : &self.comparator,
+				source : &mut self.data,
+			});
+		}
+		None
+	}
+
+	pub fn len(&self) -> usize {
+		self.data.len()
+	}
+
+	pub fn is_empty(&self) -> bool {
+		self.data.is_empty()
+	}
+
+	pub fn top(&self) -> Option<&T> {
+		if self.data.is_empty() {
+			return None;
+		}
+		Some(&self.data[0])
+	}
+}
+
+impl<T> FromIterator<T> for MinHeap<T, DefaultComparator>
+	where
+	T : Ord
+{
+	fn from_iter<I>(iter: I) -> Self
+	where
+		I: IntoIterator<Item = T>
+	{
+		let into_iter = iter.into_iter();
+		let hint_cap = match into_iter.size_hint() {
+			(lbound, Some(ubound)) => {
+				(lbound + ubound) / 2	// median
+			},
+			(lbound, None) => {
+				lbound
+			}
+		};
+		let mut result_buffer : Vec<T> = Vec::with_capacity(hint_cap);
+		for item in into_iter {
+			result_buffer.push(item);
+		}
+		MinHeap::<T, DefaultComparator>::from_vec(result_buffer, DefaultComparator)
+	}
+}
```

Comment:
```suggestion
impl<T, C> FromIterator<T> for MinHeap<T, C>
where
    C: Comparator<T> + Default,
{
    fn from_iter<I>(iter: I) -> Self
    where
        I: IntoIterator<Item = T>,
    {
        MinHeap::<T, C>::from_vec(iter.into_iter().collect(), C::default())
    }
}
```
1. DefaultComparator뿐만 아니라 모든 디폴트 trait이 있는 Comparator에 대해서 FromIterator를 구현할 수 있습니다.
2. .collect()는 해당 코드블록에서 하는 모든 일이 포함되어 있습니다. 따라서 그냥 .collect()를 하는 것이 좋습니다.

Reply:
두 제안 모두 너무나도 공감이 갑니다. 조언 감사합니다.